### ISSUE TITLE
Confusion about data processing in the FedKSeed branch.

### ISSUE THREAD

**Issue by Message00** [alibaba/FederatedScope#786]
I checked the processed training datasets from _get_loaders()_ of _utils_data.load_data_. For each sample, the _input_ids_ only includes the _Request_ part, and _labels_ only includes the _Response_ part. _zo_forward()_ will put these data into _model_ created by _AutoModelForCausalLM_ to calculate loss.

However, as far as I know, the _input_ids_ and _labels_ of _AutoModelForCausalLM_ should be the same, which includes both _Request_ and _Response_ parts. Did I overlook anything, or is this a bug?

**Comment by Message00** [comment]
@zhenqincn Hello, could you take a look at this issue? Thank you!

**Comment by zhenqincn** [comment]
Hi, if I understand correctly, the situation you're describing should only occur when the `generation` argument is set to True, which only happens when calling `model.generate()` to produce content for calculating the Rouge-L score. During model training, the data used is loaded with the `generation` argument set to False. In this case, `labels` and `input_ids` are the same, used to compute the loss.

```python

    # utils_data/llm_dataset.py
    def preprocess(self, sources, targets, tokenizer, generation):
        if generation:
            sources_tokenized, labels_tokenized = [
                self._tokenize_fn(strings, tokenizer)
                for strings in (sources, targets)
            ]
            input_ids = self._tokenize_fn(sources, tokenizer)["input_ids"]
            labels = self._tokenize_fn(targets, tokenizer)["input_ids"]
        else:
            examples = [s + t for s, t in zip(sources, targets)]
            examples_tokenized, sources_tokenized = [
                self._tokenize_fn(strings, tokenizer)
                for strings in (examples, sources)
            ]
            input_ids = examples_tokenized["input_ids"]
            labels = copy.deepcopy(input_ids)
            for label, source_len in zip(labels,
                                        sources_tokenized["input_ids_lens"]):
                label[:source_len] = DefaultToken.IGNORE_INDEX.value
        return dict(input_ids=input_ids, labels=labels)

```

**Comment by Message00** [comment]
> Hi, if I understand correctly, the situation you're describing should only occur when the `generation` argument is set to True, which only happens when calling `model.generate()` to produce content for calculating the Rouge-L score. During model training, the data used is loaded with the `generation` argument set to False. In this case, `labels` and `input_ids` are the same, used to compute the loss.
> 
> ```python
>     # utils_data/llm_dataset.py
>     def preprocess(self, sources, targets, tokenizer, generation):
>         if generation:
>             sources_tokenized, labels_tokenized = [
>                 self._tokenize_fn(strings, tokenizer)
>                 for strings in (sources, targets)
>             ]
>             input_ids = self._tokenize_fn(sources, tokenizer)["input_ids"]
>             labels = self._tokenize_fn(targets, tokenizer)["input_ids"]
>         else:
>             examples = [s + t for s, t in zip(sources, targets)]
>             examples_tokenized, sources_tokenized = [
>                 self._tokenize_fn(strings, tokenizer)
>                 for strings in (examples, sources)
>             ]
>             input_ids = examples_tokenized["input_ids"]
>             labels = copy.deepcopy(input_ids)
>             for label, source_len in zip(labels,
>                                         sources_tokenized["input_ids_lens"]):
>                 label[:source_len] = DefaultToken.IGNORE_INDEX.value
>         return dict(input_ids=input_ids, labels=labels)
> ```

Thank you for your reply. I overlook setting `args.eval_metric = 'loss'` when copy code snippets. Thank you for your code and assistance. It helps me a lot!
